### PR TITLE
EN-2465 Support MultipleSubForms in PAM forms widget EN-2414 PAM hide the form after submit in the task form widget

### PR DESCRIPTION
**Pull request by alepintus**
Hi @joewhite101 ,

- Support for MultipleSubForm added. 
- form generation fixed for nested layout components and first level fields. 
- Row and columns generation support added to show more fields in the same row.
- Entando i18n field Label generation fix when the PAM Label is not present.
- Form hidden after the submit

Could you please have a look?

Thanks!

### CODE REVIEW

**Comment by entando-service** [comment]
All is Well

**Comment by entando-service** [comment]
All is Well

**Comment by entando-service** [comment]
All is Well

File: plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/aps/system/services/kie/KieVersionTransformer.java [hunk 1]
```diff
@@ -180,7 +183,8 @@ public static KieProcessFormQueryResult pamSevenFormToPamSix(PamArray pamSeven,
 
         PamModel model = pamSeven.getModel();
         String modelName = model.getName();
-        String modelerType = "dataModelerEntry";;
+        String modelerType = "dataModelerEntry";
+        ;
```
Comment (joewhite101) [inline plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/aps/system/services/kie/KieVersionTransformer.java]: Stray semicolon

File: plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/aps/system/services/kie/KieVersionTransformer.java [hunk 2]
```diff
@@ -237,25 +241,29 @@ public static KieProcessFormQueryResult pamSevenFormToPamSix(PamArray pamSeven,
...
             Boolean addEmptyOption = field.isAddEmptyOption();
             String defaultValue = field.getDefaultValue();
             //MultipleSelector properties
-            List<String> options =  field.getListOfValues();
-         
+            List<String> options = field.getListOfValues();
+
+            Map<String, String> columnsMeta = new HashMap<>();
+
+            //MultipleSubForm Fields
```
Comment (joewhite101) [inline plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/aps/system/services/kie/KieVersionTransformer.java]: Should there be something to go with this comment?

File: plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/apsadmin/portal/specialwidget/helper/DataUXBuilder.java [hunk 3]
```diff
@@ -304,15 +322,39 @@ private T addField(KieProcessFormField field, KieFormOverride formOverride) thro
             readOnly = Boolean.parseBoolean(field.getProperty("readOnly").getValue());
         }
 
+        boolean openRow=false;
+        boolean closeRow=false;
+
+
+        if (fieldSpanInt==12) {
```
Comment (joewhite101) [inline plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/apsadmin/portal/specialwidget/helper/DataUXBuilder.java]: What is 12 here? Add a comment describing purpose

File: plugins/entando-plugin-jpkiebpm/src/main/webapp/WEB-INF/plugins/jpkiebpm/aps/jsp/widgets/bpm-datatype-form.jsp [hunk 4]
```diff
@@ -21,6 +21,21 @@
...
 <script src="<wp:resourceURL />plugins/jpkiebpm/static/js/jquery-ui.js"></script>
 <script src="<wp:resourceURL />plugins/jpkiebpm/static/js/jquery.validate.js"></script>
-<script src="<wp:resourceURL />plugins/jpkiebpm/static/js/lib/dynamic-form/jquery.dform-1.1.0.js"></script>
+
+<!-- Datatable used for MultipleSubForms fields-->
+
+<script src="<wp:resourceURL />plugins/jpkiebpm/static/js/custom-data-table.js"></script>
+<script src="<wp:resourceURL />plugins/jpkiebpm/static/js/jquery.dataTables.js"></script>
+<script src="<wp:resourceURL />plugins/jpkiebpm/static/js/dataTables.responsive.min.js"></script>
+<!--script src="<wp:resourceURL />plugins/jpkiebpm/static/js/dataTables.fixedColumns.min.js"></script>
```
Comment (joewhite101) [inline plugins/entando-plugin-jpkiebpm/src/main/webapp/WEB-INF/plugins/jpkiebpm/aps/jsp/widgets/bpm-datatype-form.jsp]: Remove if dead

File: plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/apsadmin/portal/specialwidget/helper/DataUXBuilder.java [hunk 5]
```diff
@@ -387,14 +429,33 @@ private T addField(KieProcessFormField field, KieFormOverride formOverride) thro
...
                 selectOption.add(opt);
             }
-
             multipleSelector.setOptions(selectOption);
             multipleSelector.setMaxDropdownElements(maxDropdownElements);
             multipleSelector.setMaxElementsOnTitle(maxElementsOnTitle);
             multipleSelector.setAllowClearSelection(allowClearSelection);
             multipleSelector.setAllowFilter(allowFilter);
         }
+        if (inputField instanceof MultipleSubFormField) {
```
Comment (joewhite101) [inline plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/apsadmin/portal/specialwidget/helper/DataUXBuilder.java]: This needs test coverage

File: plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/apsadmin/portal/specialwidget/helper/DataUXBuilder.java [hunk 6]
```diff
@@ -304,15 +322,39 @@ private T addField(KieProcessFormField field, KieFormOverride formOverride) thro
             readOnly = Boolean.parseBoolean(field.getProperty("readOnly").getValue());
         }
 
+        boolean openRow=false;
```
Comment (joewhite101) [inline plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/apsadmin/portal/specialwidget/helper/DataUXBuilder.java]: New block here needs test coverage

File: plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/apsadmin/portal/specialwidget/helper/DataUXBuilder.java [hunk 7]
```diff
@@ -281,13 +286,26 @@ private T addField(KieProcessFormField field, KieFormOverride formOverride) thro
...
                 inputField = (T) new InputField();
         }
-        logger.info("adding default properties to inputField");
 
         String fieldName = field.getName();
         String fieldTypeHMTL = this.typeMapping.get(field.getType());
         String fieldTypePAM = field.getType();
 
         String fieldValueExpr = this.valueMapping.get(field.getType());
+        String fieldSpan = field.getProperty("span").getValue();
```
Comment (joewhite101) [inline plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/apsadmin/portal/specialwidget/helper/DataUXBuilder.java]: Test coverage

File: plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/aps/system/services/kie/KieVersionTransformer.java [hunk 8]
```diff
@@ -462,6 +520,18 @@ private static KieProcessFormField buildField(  String id, String modelName,
             field.getProperties().add(optionsProperty);
 
         }
+        if (field.getType().equals("MultipleSubForm")) {
```
Comment (joewhite101) [inline plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/aps/system/services/kie/KieVersionTransformer.java]: Test coverage

File: plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/aps/system/services/kie/KieVersionTransformer.java [hunk 9]
```diff
@@ -283,32 +300,72 @@ public static KieProcessFormQueryResult pamSevenFormToPamSix(PamArray pamSeven,
...
                 if (components != null) {
-                    for (PamLayoutComponent component : components) {
-                        String fieldId = component.getProperties().getFieldId();
-
-                        //If the value is a form it won't be in the map. Those get ordered elsewhere. This just orders fields
-                        if (fieldMap.containsKey(fieldId)) {
-                            fieldMap.get(fieldId).setPosition(count);
-                            result.getFields().add(fieldMap.get(fieldId));
-                            count++;
+                    count = getCount(result, fieldMap, count, components, column, columnCount, columnSize);
```
Comment (joewhite101) [inline plugins/entando-plugin-jpkiebpm/src/main/java/org/entando/entando/plugins/jpkiebpm/aps/system/services/kie/KieVersionTransformer.java]: Test coverage

**Comment by entando-service** [comment]
All is Well

**Comment by entando-service** [comment]
All is Well

**Pull request merged**